jerry-core/ecma/operations/ecma-container-object.c

```diff
@@ -614,15 +647,219 @@ ecma_op_container_delete (ecma_value_t this_arg, /**< this argument */
+  {
+    return ecma_raise_type_error (ECMA_ERR_MSG ("Argument 'this' is not an iterator."));
+  }
+
+  ecma_map_object_t *map_object_p = ECMA_GET_POINTER (ecma_map_object_t,
+                                                      ext_obj_p->u.pseudo_array.u2.iterated_value_cp);
+
+  if (map_object_p == NULL)
+  {
+    return ecma_create_iter_result_object (ECMA_VALUE_UNDEFINED, ECMA_VALUE_TRUE);
+  }
+
+  ecma_object_t *internal_obj_p = ecma_get_object_from_value (map_object_p->header.u.class_prop.u.value);
+  ecma_collection_header_t *props_p = ecma_op_object_get_property_names (internal_obj_p, ECMA_LIST_NO_OPTS);
+
+  uint32_t length = props_p->item_count;
+  uint32_t index = ext_obj_p->u.pseudo_array.u1.iterator_index;
+
+  if (JERRY_UNLIKELY (index == ECMA_ITERATOR_INDEX_LIMIT))
+  {
+    /* After the ECMA_ITERATOR_INDEX_LIMIT limit is reached the [[%Iterator%NextIndex]]
+       property is stored as an internal property */
+    ecma_string_t *prop_name_p = ecma_get_magic_string (LIT_INTERNAL_MAGIC_STRING_ITERATOR_NEXT_INDEX);
+    ecma_value_t index_value = ecma_op_object_get (obj_p, prop_name_p);
+
+    if (!ecma_is_value_undefined (index_value))
+    {
+      index = (uint32_t) (ecma_get_number_from_value (index_value) + 1);
+    }
+
+    ecma_value_t put_result = ecma_op_object_put (obj_p,
+                                                  prop_name_p,
+                                                  ecma_make_uint32_value (index),
+                                                  true);
+
+    ecma_free_value (index_value);
+
+    if (!ecma_is_value_true (put_result))
+    {
+      return put_result;
+    }
+  }
+  else
+  {
+    ext_obj_p->u.pseudo_array.u1.iterator_index++;
+  }
+
+  if (index >= length)
+  {
+    ECMA_SET_POINTER (ext_obj_p->u.pseudo_array.u2.iterated_value_cp, NULL);
+    ecma_free_values_collection (props_p, 0);
+    return ecma_create_iter_result_object (ECMA_VALUE_UNDEFINED, ECMA_VALUE_TRUE);
+  }
+
+  uint8_t iterator_type = ext_obj_p->u.pseudo_array.extra_info;
+
+  ecma_value_t *ecma_value_p = ecma_collection_iterator_init (props_p);
+
+  ecma_value_t ret_value = ECMA_VALUE_UNDEFINED;
+
+  while (ecma_value_p != NULL)
+  {
+    if (index > 0)
+    {
+      index--;
+      ecma_value_p = ecma_collection_iterator_next (ecma_value_p);
+      continue;
+    }
+
+    ecma_string_t *prop_name_p = ecma_get_prop_name_from_value (*ecma_value_p);
+    ecma_property_t *property_p = ecma_find_named_property (internal_obj_p, prop_name_p);
+    JERRY_ASSERT (property_p != NULL);
+
+    if (ecma_is_value_empty (ECMA_PROPERTY_VALUE_PTR (property_p)->value))
+    {
+      ecma_value_p = ecma_collection_iterator_next (ecma_value_p);
+
+      if (ecma_value_p == NULL)
+      {
+        ret_value = ecma_create_iter_result_object (ECMA_VALUE_UNDEFINED, ECMA_VALUE_TRUE);
+      }
+      continue;
+    }
+
+    ecma_value_t value = ecma_copy_value (ECMA_PROPERTY_VALUE_PTR (property_p)->value);
+    ecma_value_t key_arg;
```

Comment:
If I read this correctly the `key_arg` is used only if the `is_set` is false. Would it be possible to only "calculate" the key_arg at the required places? Maybe move `key_arg` "load" into a method?

Reply:
In addition this same logic seems to be already used in this file.

Reply:
Thanks, I've applied your suggestion.